Lottie icons 2408 (Milestone 1)

ohrytskov opened the pull request:
#2408 

Replace Toolbar SVGs with New Static Icons (Milestone 1)

raineorshine commented on src/components/RedoIcon.tsx:
```diff
@@ -4,19 +4,43 @@ import { token } from '../../styled-system/tokens'
       fill={fill || token('colors.fg')}
-      style={style}
-      viewBox='0 -50 600 600'
+      style={{ ...style, width: `${newSize}px`, height: `${newSize}px`, marginTop: '-1px' }}
```
This -1px marginTop might be affecting the positioning across font sizes since it does not scale. You can adjust the font size using the links in the footer:
<img width="221" alt="image" src="https://github.com/user-attachments/assets/cdbf9240-ee4e-42d7-9cc7-8dc55e6294ca">

I tested the vertical centering by scrolling the toolbar a bit to the right (and I temporarily deleted the padding on the scroll arrow for closer examination).

## This PR

Font Size 18 (default):
<img width="105" alt="Brave Browser 2024-09-28 17 56 02" src="https://github.com/user-attachments/assets/08890585-6b36-42ae-99d3-8c90b22ec189">

Font Size 25:
<img width="124" alt="Brave Browser 2024-09-28 17 55 18" src="https://github.com/user-attachments/assets/c92e7c41-b74b-47bb-96a8-cb089e96f91a">

# `main`

While the icons are admittedly slightly lower than center, the difference is not amplified when the font size increases.

Font Size 18 (default):
<img width="91" alt="Brave Browser 2024-09-28 17 59 34" src="https://github.com/user-attachments/assets/3704f4ed-8602-4298-b766-e4fe15197dc9">

Font Size 25:
<img width="114" alt="Brave Browser 2024-09-28 17 59 48" src="https://github.com/user-attachments/assets/68b9d25a-f35a-43fb-b602-2ac264e04336">

Note that I'm not sure if it is the toolbar button icons or the arrow icon is wrong, but either way they need to be aligned vertically along their centers regardless of font size.

Let me know if that makes sense!

ohrytskov replied:
The new icons are centered within the SVG and include some padding, while the previous icons were aligned to the top-left corner. I aimed to position the new icons at the same level as the previous ones for consistency. Scaling with state.fontSize is part of Milestone 3, so for now, we're focused on achieving perfect positioning without scaling. A good option for now would be to remove the margin-top: -1px entirely, as all the icons will be aligned on the same level after replacement.

raineorshine replied:
Great! We'll revisit in Milestone 3.

raineorshine commented on src/components/icons/IndentIcon.tsx:
```diff
@@ -0,0 +1,89 @@
+import { token } from '../../../styled-system/tokens'
+import IconType from '../../@types/Icon'
+
+// eslint-disable-next-line jsdoc/require-jsdoc, react-refresh/only-export-components
```
I'm not sure why `react-refresh/only-export-components` is needed here, since it is a component. does it need to be typed as `FC`?

ohrytskov replied:
The eslint-disable comment was carried over from another file where it was needed, but it's redundant in this component and can be removed.

raineorshine commented on src/components/RedoIcon.tsx:
```diff
@@ -4,19 +4,43 @@ import { token } from '../../styled-system/tokens'
 /** A redo icon. */
-const RedoIcon = ({ fill, size = 18, style, className }: Icon) => {
+const RedoIcon = ({ fill, size = 14, style = {}, className }: Icon) => {
+  const scalingFactor = 1.37
```
If the scaling factor is the same for all icons, we should probably add it to `constants.ts` rather than repeating it each time.

ohrytskov replied:
Based on the current set of replaced SVGs, the scaling factor is indeed the same for all icons, though this wasn't clear before the replacements, and it’s still unknown for other icons. We can move the scaling factor to constants.ts, and if a different factor is needed for a specific icon later on, it can be overridden by defining a new one instead of importing from constants. So yes, let's move it to constants.

ohrytskov replied:
Scaling factor has been moved to constants.

raineorshine commented on src/components/RedoIcon.tsx:
```diff
@@ -4,19 +4,43 @@ import { token } from '../../styled-system/tokens'
 
 /** A redo icon. */
-const RedoIcon = ({ fill, size = 18, style, className }: Icon) => {
+const RedoIcon = ({ fill, size = 14, style = {}, className }: Icon) => {
```
I think changing the default font size to 14 is going to break some of the icons that are used in multiple places, such as https://github.com/cybersemics/em/blob/main/src/components/icons/HomeIcon.tsx. The size should correspond to font size, which defaults to 18. Can you change the default size to 18 and the scale factor to 1.066, or am I missing something?

Can you change the default font size to 18 and the scale factor to 1.066, or does that mess things up?

ohrytskov replied:
Scaling is a part of Milestone 3, so we can keep all default values of size for now.

Factors to Keep in Mind for Milestone 3:

Size can be set by passing the parameter when rendering the icon component.
The default font size is set to 18 in constants, and it will remain as 18, while each icon component has its own default size prop.
The HomeIcon currently doesn’t have a default size.
Undo and redo are used not only in the toolbar but also in the modal window. We can pass size={14} explicitly from the modal window component for these icons.

raineorshine replied:
Thanks! We can revisit in Milestone 3.

raineorshine commented:
Note that the puppeteer snapshot tests are expected to fail because the toolbar appearance is changing. You can regenerate all the snapshots at the end by running `yarn test:puppeteer -u`.

trevinhofmann commented on src/components/icons/OutdentIcon.tsx:
```diff
@@ -0,0 +1,88 @@
+import { token } from '../../../styled-system/tokens'
+import IconType from '../../@types/Icon'
+
+// eslint-disable-next-line jsdoc/require-jsdoc, react-refresh/only-export-components
```
The `react-refresh/only-export-components` rule can be re-enabled safely, and we should be able to remove `jsdoc/require-jsdoc` by adding a simple JSDoc comment here.

trevinhofmann commented on src/components/icons/Sort.tsx:
```diff
@@ -11,56 +12,166 @@ import head from '../../util/head'
 
-// eslint-disable-next-line jsdoc/require-jsdoc, react-refresh/only-export-components
-const Icon = ({ size = 20, style }: IconType) => {
+/** Main Icon Component. */
```
```suggestion
/** Sort Icon Component. */
```

trevinhofmann commented on src/components/icons/StarIcon.tsx:
```diff
@@ -2,26 +2,41 @@ import { icon } from '../../../styled-system/recipes'
-const Icon = ({ fill, size = 20, style }: IconType) => {
+/**
+ *
+ */
```
This should be `/** Favorite Icon Component. */` rather than left blank.

trevinhofmann reviewed (changes requested):
Thank you, @ohrytskov! I've left a few initial comments. Also, the snapshots still need to be updated with `yarn test:puppeteer -u`.

raineorshine commented on src/components/icons/subCategorizeOneIcon.tsx:
```diff
@@ -0,0 +1,60 @@
+const subCategorizeOneIcon = ({ fill = 'black', size = 20, style = {}, className }: IconType) => {
+  const scalingFactor = 1.37
+  const newSize = size * scalingFactor
+  const strokeColor = style.fill || fill || token('colors.fg')
```
Yes, I'm open to that. Let's wait till after Milestone 2 of #2408 so that we can see the full extent of the common icon functionality including static styles and animation logic.

raineorshine commented on src/components/icons/DeleteIcon.tsx:
```diff
@@ -1,23 +1,73 @@
+import IconType from '../../@types/Icon'
+
+/** Delete icon. */
+const DeleteIcon = ({ fill, size = 20, style = {}, className }: IconType) => {
```
Yes, we'll have to normalize these sizes, but it's okay to leave it to Milestone 3 as long as the visual is correct now.

raineorshine commented on src/components/icons/IndentIcon.tsx:
```diff
@@ -0,0 +1,90 @@
+/**
+ *
+ */
+const IndentIcon = ({ fill, size = 14, style = {}, className }: IconType) => {
```
Ideally all the icons will be scaled the same and will default to size 18. Plan on addressing that in Milestone 3.

trevinhofmann commented:
Hi, @ohrytskov. I see you've added commits since my last review. When the PR is ready again for review, please just assign it back to me and/or let me know in the comments. Thank you!

ohrytskov commented:
Hi @trevinhofmann, thank you for your initial feedback. I've made the updates to the PR. Could you please take another look?

raineorshine commented:
@ohrytskov Could you update the puppeteer snapshot tests, too?

ohrytskov commented:
@raineorshine Thanks for the reminder! I ran yarn test:puppeteer -u, and the snapshots are all updated now.

trevinhofmann reviewed (changes requested):
Thank you, @ohrytskov! There are a few merge conflicts with the Puppeteer snapshots which will need to be resolved.

ohrytskov commented:
Hi @trevinhofmann,

In this update:

1. Renamed the scaling factor and updated its description for clarity.
2. Added ConvertToNoteIcon for the Convert to Note action, while keeping PencilIcon for the Note action.

Please review and let me know if any changes are needed.

Thanks!

trevinhofmann reviewed (changes requested):
Thank you very much for making those changes, @ohrytskov! Could you please resolve the merge conflicts before I give this a final glance?

ohrytskov commented:
The merge conflicts have been resolved, and the branch is now ready for final review.

raineorshine commented:
Thanks. Please pause work on this until #2432 is merged. Expecting significant merge conflicts as previously discussed.

ohrytskov commented:
Got it, @raineorshine. I’ll pause work on this branch until #2432 is merged. I’ll keep an eye out for potential merge conflicts once that’s done. Thanks for the update!

raineorshine commented:
Several snapshot tests seem to be failing, but I'm not clear why. Browserless should get rid of platform differences, so the same snapshots are generated regardless of your machine.

- font-size-22-one-thought-1-diff - React error I haven't seen before in a test.
- font-size-22-initial-load-1-diff - Related to the Swap Parent toolbar icon.
- drag-drop-hover-after-table-column-one-1-diff - Appears to be an alert fade-in timing issue (Animations will be disabled in tests soon: #2110)

raineorshine commented:
I'm going to let @trevinhofmann review and help you fix the snapshot tests.

raineorshine reviewed (changes requested):
Looks good, thanks!

Regarding the timeout comments, I'd like to explain where I'm coming from. I understand the the tests can be flaky sometimes, which is an annoying problem. We are continually trying to make the tests more stable across platforms. `docker` and `browserless` have helped a lot in ensuring tests are run the same and snapshots match across platforms. Different machines will still run tests at different speeds, which can sometimes affect test outcomes if the timeouts are too short.

That said, we need to understand the implications of increasing sleeps in the tests. This slows down tests for *every developer that works on the project*. It has a major impact on DX which should not be underestimated. This is why I created #2110, #2111, and #2054. We're currently in the middle of a big effort to speed up the puppeteer tests. This is especially important as the Puppeteer tests are the most realistic integration tests. They are very important for catching real-world usage and browser-specific behaviors, and they will only grow in number.

Increasing a delay from 100ms to 1000ms may seem like a small change, but it adds up and affects the whole team. For that reason I will be defending the current timeouts and requiring substantial explanation and justification for increasing them. If it's absolutely necessary, that I'm totally okay with that. But we need to prove that it is absolutely necessary and have good documentation for it given the repercussions.

Hope that makes sense, and I have communicated the need without coming across as too heavy-handed :).

raineorshine commented:
You probably noticed that the tests were refactored today... twice! They are stable now. The new technique is to import the `page` object directly from setup. Let me know if you need any help.

ohrytskov commented on src/e2e/puppeteer/__tests__/drag-and-drop.ts:
```diff
@@ -239,7 +239,7 @@ describe('drop', () => {
 
     await dragAndDropThought('a', 'd', { position: 'after', mouseUp: true })
-
+    await sleep(1000)
```
The 1000ms delay in the drag-and-drop tests was introduced solely for debugging purposes.

trevinhofmann replied:
Safely removed in https://github.com/cybersemics/em/pull/2429/commits/88322dc8adb2d78e113f4654a672a1956dc8a264

ohrytskov commented on src/shortcuts/archive.tsx:
```diff
@@ -1,11 +1,10 @@
 import { archiveThoughtActionCreator as archiveThought } from '../actions/archiveThought'
 import { errorActionCreator as error } from '../actions/error'
+import ArchiveIcon from '../components/icons/ArchiveIcon'
+// This line was omitted and should be included
```
Comment removed as suggested.

ohrytskov commented on src/e2e/puppeteer/helpers/waitForEditable.ts:
```diff
@@ -7,7 +7,7 @@ interface Options {
  * Wait for editable div that contains the given value.
  */
-const waitForEditable = async (page: Page, value: string, { timeout }: Options = { timeout: 6000 }) =>
+const waitForEditable = async (page: Page, value: string, { timeout }: Options = { timeout: 60000 }) =>
```
The increased timeout was added temporarily for debugging purposes to investigate a specific issue. I don't anticipate needing more than 6 seconds in normal scenarios.

raineorshine replied:
Got it! Sounds good.

Just make sure to add a comment on the PR for changes that you are still testing, or switch the PR to a draft. Otherwise we will assume that it is ready for review.

ohrytskov commented on src/e2e/puppeteer/__tests__/ContextView.ts:
```diff
@@ -27,14 +27,16 @@ it('load buffered ancestors of contexts when context view is activated', async (
+  await sleep(1000)
 
   await clickThought('m')
+  await scrollIntoView('[data-testid="toolbar-icon"][aria-label="Settings"]')
```
The reference to the Settings button was part of the debugging process to figure out why the test wasn't working. The main idea behind using scrollIntoView was to ensure the Context View button was visible and clickable. Without scrolling, buttons outside the viewport can't be clicked, due to security and optimization limitations.

ohrytskov replied:
Scrolling directly to the Context View button left it only partially visible, causing click failures. To address this, initial scrolling was adjusted to the Settings button to better center the Context View. Following optimization suggestions, the scroll now targets the next sibling, the Note button, for improved visibility and click success.

trevinhofmann replied:
This explanation made sense to me, but I went ahead and tested to verify. Switching from `Note` to `Context View` does indeed create an issue with clicking the icon.

* https://github.com/cybersemics/em/pull/2429/commits/8956f24bd3268892471e1c2c6a68224924d3f081
* https://github.com/cybersemics/em/pull/2429/commits/c9d9165bb4666c1bf1587f14e4bdccd8a4dd0aa4

ohrytskov commented on src/e2e/puppeteer/helpers/scrollIntoView.ts:
```diff
@@ -0,0 +1,13 @@
+  await page.evaluate(selector => {
+    const element = document.querySelector(selector)
+    if (element) {
+      element.scrollIntoView({ behavior: 'smooth', block: 'start' })
```
The scrolling behavior has been updated to 'instant' to improve test speed. Additionally, the 1-second sleep delay (used after scrolling and before the Context View click) has been removed, while the other two sleep calls have been reverted to their original durations. These changes support our ongoing effort to minimize animations in Puppeteer tests.

ohrytskov commented on src/e2e/puppeteer/helpers/scrollIntoView.ts:
```diff
@@ -0,0 +1,13 @@
+const scrollIntoView = async (page: Page, selector: string) => {
+  await page.evaluate(selector => {
+    const element = document.querySelector(selector)
+    if (element) {
```
Got it! And thank you for pointing this out. I have modified the function to throw an explicit error if the element isn't found, ensuring that any absence is caught early in the testing process.

trevinhofmann reviewed (approved):
Thank you for your patience with merge conflicts and test flakiness, @ohrytskov!

All of the new timeouts/sleeps seem to have been cleaned up, @raineorshine, and tests are passing reliably (in the CI; I have occasional flakiness on my own machine, unrelated to these changes).

This PR looks good to me at this point 🎉

raineorshine commented on src/e2e/puppeteer/__tests__/ContextView.ts:
```diff
@@ -33,6 +34,8 @@ it('load buffered ancestors of contexts when context view is activated', async (
   await sleep(100)
 
   await clickThought('m')
+  await scrollIntoView('[data-testid="toolbar-icon"][aria-label="Note"]')
```
Scrolling to the Note button is no better than scrolling to Settings button. You are still tightly coupling the Context View test to an unrelated button. What happens if the buttons change position in the Toolbar? Then this test will break. Instead you need to:

1. Scroll to the Context View button plus a few pixels.
2. Scroll to the "next" button after the Context View regardless of what that button is (using DOM traversal).

raineorshine commented on src/components/RedoIcon.tsx:
```diff
@@ -2,21 +2,44 @@ import { css, cx } from '../../styled-system/css'
       fill={fill || token('colors.fg')}
-      style={style}
-      viewBox='0 -50 600 600'
+      style={{ ...style, width: `${newSize}px`, height: `${newSize}px` }}
```
Why set `width` and `height` in the `style` attribute when they are already set in the width and height attributes? It seems like it should be either one or the other.

raineorshine commented on src/shortcuts/archive.tsx:
```diff
@@ -33,54 +31,27 @@ const exec: Shortcut['exec'] = (dispatch, getState, e) => {
     } else if (noteFocus) {
       const path = state.cursor || HOME_PATH
       const childNote = findAnyChild(state, head(path), child => child.value === '=note')
-      // we know there is a =note child if noteFocus is true
-      // we just need to get the Child object so that archiveThought has the full path
       const pathNote = appendToPath(path, childNote!.id)
       dispatch(archiveThought({ path: pathNote }))
     } else {
-      // clear the undo alert timer to prevent previously cleared undo alert from closing this one
       clearTimeout(undoArchiveTimer)
-
-      // close the alert after a delay
-      // only close the alert if it is a ThoughtArchive alert
```
Not sure why these comments were deleted?

raineorshine commented on src/components/icons/splitSentencesIcon.tsx:
```diff
@@ -0,0 +1,77 @@
+import { ICON_SCALING_FACTOR } from '../../constants'
+
+/** SplitSentences icon. */
+const splitSentencesIcon = ({ fill = 'black', size = 20, style = {}, cssRaw }: IconType) => {
```
Component names should be PascalCase

raineorshine commented on src/components/icons/subCategorizeAllIcon.tsx:
```diff
@@ -0,0 +1,74 @@
+import { ICON_SCALING_FACTOR } from '../../constants'
+
+/** SubCategorize All icon. */
+const subCategorizeAllIcon = ({ fill = 'black', size = 20, style = {}, cssRaw }: IconType) => {
```
Component names should be PascalCase

raineorshine commented on src/components/icons/subCategorizeOneIcon.tsx:
```diff
@@ -0,0 +1,61 @@
+import { ICON_SCALING_FACTOR } from '../../constants'
+
+/** Sub Categorize One icon. */
+const subCategorizeOneIcon = ({ fill = 'black', size = 20, style = {}, cssRaw }: IconType) => {
```
Component names should be PascalCase

raineorshine commented on src/shortcuts/home.tsx:
```diff
@@ -1,36 +1,15 @@
-    </g>
-  </svg>
-)
+import HomeToolbarIcon from '../components/icons/HomeToolbarIcon'
```
At some point `HomeIcon` and `HomeToolbarIcon` will need to be combined.

raineorshine commented on src/components/icons/SettingsIcon.tsx:
```diff
@@ -1,25 +1,41 @@
 import IconType from '../../@types/Icon'
+import { ICON_SCALING_FACTOR } from '../../constants'
+
+/** Settings icon refactored. */
```
Let's just document this component as "Settings icon."

raineorshine commented on src/components/icons/ShareIcon.tsx:
```diff
@@ -2,21 +2,48 @@ import { css, cx } from '../../../styled-system/css'
 import IconType from '../../@types/Icon'
+import { ICON_SCALING_FACTOR } from '../../constants'
+
+/** Share icon refactored. */
```
Let's just document this component as "Share icon."

raineorshine commented on src/components/icons/Sort.tsx:
```diff
@@ -1,66 +1,176 @@
-    </g>
-  </svg>
-)
+/** Ascending Icon Component refactored. */
```
Remove "refactored"

raineorshine commented on src/components/icons/Sort.tsx:
```diff
@@ -1,66 +1,176 @@
+  )
+}
+
+/** Descending Icon Component refactored. */
```
Remove "refactored"

raineorshine commented on src/components/icons/StarIcon.tsx:
```diff
@@ -2,28 +2,33 @@ import { css, cx } from '../../../styled-system/css'
 import IconType from '../../@types/Icon'
+import { ICON_SCALING_FACTOR } from '../../constants'
+
+/** Favorite icon refactored. */
```
Remove "refactored"

raineorshine commented on src/components/icons/StarIcon.tsx:
```diff
@@ -2,28 +2,33 @@ import { css, cx } from '../../../styled-system/css'
+import { ICON_SCALING_FACTOR } from '../../constants'
+
+/** Favorite icon refactored. */
+const FavoriteIcon = ({ fill, size = 14, style = {}, cssRaw }: IconType) => {
```
The file name should always match the component name, so rename this file to `FavoritesIcon.tsx`.

raineorshine commented on src/components/icons/StrikethroughIcon.tsx:
```diff
@@ -1,26 +1,43 @@
+import IconType from '../../@types/Icon'
+import { ICON_SCALING_FACTOR } from '../../constants'
+
+/** Strikethrough icon refactored. */
```
Remove "refactored"

raineorshine commented on src/components/icons/SwapParentIcon.tsx:
```diff
@@ -1,24 +1,36 @@
+import IconType from '../../@types/Icon'
+import { ICON_SCALING_FACTOR } from '../../constants'
+
+/** SwapParent icon refactored. */
```
Remove "refactored"

raineorshine commented on src/components/icons/TextColor.tsx:
```diff
@@ -1,43 +1,31 @@
 import IconType from '../../@types/Icon'
+import { ICON_SCALING_FACTOR } from '../../constants'
+
+/** TextColor Icon refactored. */
```
Remove "refactored"

raineorshine commented on src/components/icons/UnderlineIcon.tsx:
```diff
@@ -1,28 +1,35 @@
-    <path d='M230,270H70c-8.284,0-15,6.716-15,15s6.716,15,15,15h160c8.284,0,15-6.716,15-15S238.284,270,230,270z'></path>
-  </svg>
-)
+/** Underline icon refactored. */
```
Remove "refactored"

raineorshine reviewed (changes requested):
Think of JSDOC comments as permanent documentation for a function or variable. They are good for describing what a function does, why it does it, where it's used, and any edge cases that the developer should know about. This is why I recommend removing "refactored" from the JSDOC's. It's not pertinent information for future developers. We always have the commit history to see how a piece of code has been refactored over time.

I noticed a positioning issue with the icons that are used in the QuickDrop components:

## Steps to Reproduce

Drag a thought.

## Current Behavior

The Copy to Clipboard icon smaller than the other icons and is not centered.

<img width="110" alt="Brave Browser 2024-10-15 08 35 43" src="https://github.com/user-attachments/assets/ac6c27bb-c34e-4151-9910-4a220251db8f">

## Expected Behavior

QuickDrop icons should be the same size and centered.

<img width="103" alt="Brave Browser 2024-10-15 08 48 25" src="https://github.com/user-attachments/assets/d2f47948-3145-444c-ba94-eaf2fdadbe8f">

ohrytskov commented:
Recent Updates Ready for Review:

SVG Update: Removed width and height from top-level SVG tags.
New Helper: Added scrollBy in ContextView test.
Renaming: Changed StarIcon.tsx to FavoritesIcon.tsx.
Consistency: Standardized icon component names and enhanced the Copy Clipboard icon.
Docs: Updated comments in icons and archive shortcut.
